### PR TITLE
Add moving statement and declaration action

### PR DESCRIPTION
**Pull request by grenki**

### CODE REVIEW

File: src/com/goide/editor/GoStatementMover.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.util.PsiTreeUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+
+public class GoStatementMover extends LineMover {
+  @Override
+  public boolean checkAvailable(@NotNull Editor editor, @NotNull PsiFile file, @NotNull MoveInfo info, boolean down) {
+    if (!(file instanceof GoFile || super.checkAvailable(editor, file, info, down))) return false;
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: why `super.checkAvailable(editor, file, info, down)` is more important?

File: src/com/goide/editor/GoStatementMover.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      return true;
+    }
+    if (commonParent instanceof GoImportList) {
+      GoImportDeclaration toMove2 =
+        (GoImportDeclaration)getElementOfTypeAfter(commonParent, elementRange, GoImportDeclaration.class, down);
+      info.toMove2 = toMove2 == null ? null : new LineRange(toMove2);
+      return true;
+    }
+    setUpInfo(info, elementRange, commonParent, down);
+    return true;
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: So you don't give a chance any other line mover, why?

File: src/com/goide/editor/GoStatementMover.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  @Nullable
+  private static Pair<PsiElement, PsiElement> getLogicElementRange(@NotNull Pair<PsiElement, PsiElement> elementRange,
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: Use Couple instead Pair, I believe IDEA should highlight is with warning

File: src/com/goide/editor/GoStatementMover.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      }
+      if (range.getStartOffset() >= start && range.getStartOffset() < end && range.getEndOffset() >= end) {
+        endElement = element;
+      }
+    }
+    if (!startElement.getParent().isEquivalentTo(endElement.getParent())) {
+      return null;
+    }
+
+    return new Pair<>(startElement, endElement);
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: `Pair.create`

File: src/com/goide/editor/GoStatementMover.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      PsiTreeUtil.findCommonParent(primeElementRange.first, primeElementRange.second);
+    if (commonParent == null) return false;
+
+    Pair<PsiElement, PsiElement> elementRange = getLogicElementRange(primeElementRange, commonParent);
+    if (elementRange == null) return false;
+    if (commonParent == elementRange.first) commonParent = commonParent.getParent();
+    info.toMove = new LineRange(elementRange.first, elementRange.second);
+    if (elementRange.first instanceof GoTopLevelDeclaration) {
+      if (!file.isValid()) {
+        return true;
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: Why do you checking the file?
Why do you tell platform that your line mover is an appropriate for invalid file?

**Comment by zolotov** [comment]
Works bad for
```
package main
import (
	"i<caret>o"
)
```

**Comment by zolotov** [comment]
Works bad on anonymous functions

```
package main

import "fmt"

func main() {
	var _ func() = fu<caret>nc() {
	}
}
```

```
package main

func main() {
	A(fu<caret>nc() {
	})
}

func A(func()) {
	
}
```

File: src/com/goide/editor/GoStatementMover.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+
+public class GoStatementMover extends LineMover {
+  @Override
+  public boolean checkAvailable(@NotNull Editor editor, @NotNull PsiFile file, @NotNull MoveInfo info, boolean down) {
+    if (!(file instanceof GoFile && super.checkAvailable(editor, file, info, down))) return false;
+
+    Pair<PsiElement, PsiElement> primeElementRangePair = getElementRange(editor, file, getLineRangeFromSelection(editor));
+    if (primeElementRangePair == null) return false;
+    Couple<PsiElement> primeElementRange = new Couple<PsiElement>(primeElementRangePair.first, primeElementRangePair.second);
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: Couple.of

File: src/com/goide/editor/GoStatementMover.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    return new Couple<PsiElement>(startElement, endElement);
+  }
+
+  private static boolean setUpInfo(@NotNull MoveInfo info,
+                                   @NotNull Couple<PsiElement> range,
+                                   @NotNull PsiElement parent,
+                                   boolean down) {
+    info.toMove = new LineRange(range.first, range.second);
+    info.toMove2 = null;
+    GoTopLevelDeclaration block = PsiTreeUtil.getParentOfType(parent, GoTopLevelDeclaration.class);
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: bad name

File: src/com/goide/editor/GoStatementMover.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      }
+      if (range.getStartOffset() >= start && range.getStartOffset() < end && range.getEndOffset() >= end) {
+        endElement = element;
+      }
+    }
+    if (!startElement.getParent().isEquivalentTo(endElement.getParent())) {
+      return null;
+    }
+
+    return new Couple<PsiElement>(startElement, endElement);
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: Couple.of

File: src/com/goide/editor/GoStatementMover.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    if (range.getStartOffset() == start && (children.length == 0 || children[0].getTextOffset() > start) ||
+        range.getEndOffset() == end && (children.length == 0 || children[children.length - 1].getTextRange().getEndOffset() < end)) {
+      return new Couple<PsiElement>(parent, parent);
+    }
+
+    PsiElement startElement = elementRange.first;
+    PsiElement endElement = elementRange.second;
+    for (PsiElement element : children) {
+      range = element.getTextRange();
+      if (range.getStartOffset() <= start && range.getEndOffset() <= end && range.getEndOffset() > start) {
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: the explanation is needed, at least extract variable with sensible name

File: src/com/goide/editor/GoStatementMover.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      }
+    }
+    if (!startElement.getParent().isEquivalentTo(endElement.getParent())) {
+      return null;
+    }
+
+    return new Couple<PsiElement>(startElement, endElement);
+  }
+
+  private static boolean setUpInfo(@NotNull MoveInfo info,
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: ok, imagine I want to use this method. I see `parent` parameter. What is it? Parent of what? What should I pass here?

File: src/com/goide/editor/GoStatementMover.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      ? new LineRange(nearLine, nearLine + 1)
+      : null;
+    return true;
+  }
+
+  @Nullable
+  private static PsiElement getElementOfTypeAfter(@NotNull PsiElement parent,
+                                                  @NotNull Couple<PsiElement> range,
+                                                  @NotNull Class<? extends PsiElement> clazz,
+                                                  boolean after) {
```
Comment (zolotov) [inline src/com/goide/editor/GoStatementMover.java]: Whose parent? What does `after` mean? How couple of element fit to `getElementOfTypeAfter` name?

File: tests/com/goide/editor/GoStatementMoverTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    else {
+      myFixture.performEditorAction(IdeActions.ACTION_MOVE_STATEMENT_UP_ACTION);
+      myFixture.checkResultByFile(getBasePath() + "/" + testName + "-afterUp.go", true);
+    }
+  }
+
+  @NotNull
+  @Override
+  public String getTestDataPath() {
+    return "testData/";
```
Comment (zolotov) [inline tests/com/goide/editor/GoStatementMoverTest.java]: ?

File: src/com/goide/editor/GoStatementMover.java (hunk 9 above)
Comment (grenki) [inline src/com/goide/editor/GoStatementMover.java]: Add explanation in comment
    // get range which contains TextRange(start, end) of top level elements
    // common parent of elements is straight parent for each element

**Comment by zolotov** [comment]
merged manually

**Pull request closed without merging**